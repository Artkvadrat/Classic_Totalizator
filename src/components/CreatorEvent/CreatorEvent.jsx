/* eslint-disable camelcase */
import React, { useEffect } from 'react';
import { Skeleton } from 'antd';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import moment from 'moment';

import {
  loadData,
  createEvent,
  changeFieldEvent
} from '../../ducks/creatorEvent/creatorEvent';
import selector from './selector/CreatorEvent.selector';
import styles from './CreatorEvent.module.css';

const CreatorEvent = () => {
  const { participants, sports, addEvent, isLoading } = useSelector(selector);
  const { participant_Id1, participant_Id2, startTime, margin } = addEvent;
  const dispatch = useDispatch();
  const history = useHistory();

  useEffect(() => {
    dispatch(loadData());
  }, [dispatch]);

  const handleChange = (event) => {
    const fieldName = event.target.name;
    const fieldValue =
      event.target.name === 'margin' ? +event.target.value : event.target.value;

    dispatch(changeFieldEvent(fieldName, fieldValue));
  };

  const submit = (e) => {
    e.preventDefault();
    let isFullForm = false;
    let event = {
      ...addEvent
    };
    if (participant_Id1 === '' && participants.length > 0) {
      event = {
        ...event,
        participant_Id1: participants[0].id
      };
    }
    if (participant_Id2 === '' && participants.length > 0) {
      event = {
        ...event,
        participant_Id2: participants[0].id
      };
    }
    if (
      +margin > 0 &&
      participant_Id1 !== participant_Id2 &&
      new Date(startTime) > Date.now()
    ) {
      isFullForm = true;
    }
    if (isFullForm) {
      dispatch(
        createEvent({ ...event, startTime: moment(startTime).format() })
      );
      history.push('/');
    }
  };

  return isLoading ? (
    <Skeleton active />
  ) : (
    <form className={styles['form-creator']} onSubmit={submit}>
      <span className={styles['sport-name']}>{sports[0].name}</span>

      <span>Select the first participant of event </span>
      <select
        className={styles['player-selector']}
        name="participant_Id1"
        onChange={handleChange}
        value={participant_Id1}
        onBlur={undefined}
      >
        {participants.map(({ id, name }) => (
          <option className={styles['player1-option']} key={id} value={id}>
            {name}
          </option>
        ))}
      </select>

      <span>Select the second participant of event </span>
      <select
        className={styles['player-selector']}
        name="participant_Id2"
        value={participant_Id2}
        onChange={handleChange}
        onBlur={undefined}
      >
        {participants.map(({ id, name }) => (
          <option className={styles['player2-option']} key={id} value={id}>
            {name}
          </option>
        ))}
      </select>

      <span>Enter the margin, %</span>
      <input
        className={styles['margin-value']}
        type="number"
        name="margin"
        value={margin}
        onChange={handleChange}
      />

      <span>Enter the date of event</span>
      <input
        className={styles['start-data']}
        type="datetime-local"
        name="startTime"
        value={startTime}
        required
        onChange={handleChange}
      />

      <input
        className={styles['create-action']}
        type="submit"
        value="Create new event"
      />
    </form>
  );
};
export default CreatorEvent;
