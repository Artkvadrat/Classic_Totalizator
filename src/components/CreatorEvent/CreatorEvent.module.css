.form-creator {
    max-width: 500px;
    margin: 50px auto;
    padding: 15px;
    border: 2px solid #EEECEC;
}
input {
    outline: none;
    padding: 5px;
    border-radius: 4px;
    border: 1px solid #EEECEC;
}
span,
.sport-name,
.margin-value,
.create-action {
    display: block;
}
.sport-name {
    text-align: center;
    margin-bottom: 10px;
    font-size: 25px;
    font-weight: 700;
}

.outcome-selector,
.player-selector {
    min-width: 180px;
    border-radius: 4px;
    outline: none;
    padding: 5px;
    margin: 5px 0 10px;
}
.margin-value,
.start-data,
.start-time {
    margin: 5px 0 10px;
}

.start-data {
    margin-right: 10px;
}

.create-action {
    min-width: 150px;
    margin: 35px auto 0;
    background-color: #1890ff;
    cursor: pointer;
    color: #fff;
    border: none;
}
.create-action:hover {
    background-color: #2995fd;
}
.possible-results {
    display: flex;
    flex-wrap: wrap;
    margin-bottom: 10px;
}
.possible-results span{
    margin: 5px 10px 0 0;
    align-items: center;
}

.player1-option,
.player2-option {}