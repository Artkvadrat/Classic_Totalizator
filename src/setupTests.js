import enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

enzyme.configure({ adapter: new Adapter() });

const localStorageMock = {
  getItem: jest.fn(),
  setItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn()
};
window.localStorage = localStorageMock;
