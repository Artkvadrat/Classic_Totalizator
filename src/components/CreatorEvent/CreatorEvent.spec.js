import React from 'react';
import { shallow, mount } from 'enzyme';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import {
  loadData,
  createEvent,
  changeFieldEvent
} from '../../ducks/creatorEvent/creatorEvent';

import '../../setupTests';

import CreatorEvent from './CreatorEvent';

jest.mock('react-router-dom', () => ({
  useHistory: jest.fn()
}));

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn()
}));

jest.mock('../../ducks/creatorEvent/creatorEvent', () => ({
  loadData: jest.fn(),
  createEvent: jest.fn(),
  changeFieldEvent: jest.fn(),
  clearData: jest.fn()
}));

jest.mock('moment', () => () => ({
  format: () => new Date()
}));

describe('CreatorEvent component', () => {
  const history = {
    push: jest.fn()
  };
  const dispatch = jest.fn(() => Promise.resolve());

  beforeEach(() => {
    useDispatch.mockReturnValue(dispatch);
    useHistory.mockReturnValue(history);
    loadData.mockReturnValue(Symbol.for('load'));
    changeFieldEvent.mockReturnValue(Symbol.for('change'));
    createEvent.mockReturnValue(Symbol.for('create'));
    jest.clearAllMocks();

    useSelector.mockReturnValue({
      participants: [
        { id: '2', name: 'vlad' },
        { id: '3', name: 'vova' }
      ],
      sports: [{ id: 1, name: 'UFC' }],
      addEvent: {
        participant_Id1: '2',
        participant_Id2: '3',
        startTime: '2021-05-23T05:50:49',
        margin: 5
      },
      isLoading: false
    });
  });

  it('should dispatch  action on mount', () => {
    expect(dispatch).not.toHaveBeenCalled();
    expect(loadData).not.toHaveBeenCalled();

    mount(<CreatorEvent />);

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('load'));
    expect(loadData).toHaveBeenCalledTimes(1);
  });

  it('should render form', () => {
    const result = shallow(<CreatorEvent />);
    const form = result.find('.form-creator');
    expect(form.exists()).toEqual(true);
    expect(result).toMatchSnapshot();
  });

  it('should render Skeleton without', () => {
    useSelector.mockReturnValue({
      participants: [
        { id: '2', name: 'vlad' },
        { id: '3', name: 'vova' }
      ],
      sports: [{ id: 1, name: 'UFC' }],
      addEvent: {
        participant_Id1: '2',
        participant_Id2: '3',
        startTime: '2021-05-23T05:50:49',
        margin: 5
      },
      isLoading: true
    });
    const result = shallow(<CreatorEvent />);
    expect(result).toMatchSnapshot();
  });

  it('should create event on submit form', () => {
    const result = shallow(<CreatorEvent />);
    const fakeEvent = {
      preventDefault: jest.fn()
    };

    const form = result.find('.form-creator');
    form.simulate('submit', fakeEvent);

    expect(fakeEvent.preventDefault).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('create'));
    expect(createEvent).toHaveBeenCalledTimes(1);
    expect(history.push).toHaveBeenCalledTimes(1);
  });

  it("shouldn't create event on submit form", () => {
    useSelector.mockReturnValue({
      participants: [
        { id: '2', name: 'vlad' },
        { id: '3', name: 'vova' }
      ],
      sports: [{ id: 1, name: 'UFC' }],
      addEvent: {
        participant_Id1: '',
        participant_Id2: '',
        startTime: '2021-05-23T05:50:49',
        margin: 5
      },
      isLoading: false
    });

    const result = shallow(<CreatorEvent />);
    const fakeEvent = {
      preventDefault: jest.fn()
    };
    const form = result.find('.form-creator');
    form.simulate('submit', fakeEvent);

    expect(fakeEvent.preventDefault).toHaveBeenCalled();
    expect(dispatch).not.toHaveBeenCalled();
    expect(createEvent).not.toHaveBeenCalled();
    expect(history.push).not.toHaveBeenCalled();
  });
  it('should dispatch changeFieldEvent action on change field form', () => {
    const result = shallow(<CreatorEvent />);
    const fakeEventMargin = {
      target: {
        name: 'margin',
        value: '4'
      }
    };
    const fakeEventDate = {
      target: {
        name: 'startTime',
        value: '2021-05-23T11:18:21'
      }
    };

    const margin = result.find('.margin-value');
    margin.simulate('change', fakeEventMargin);
    const date = result.find('.start-data');
    date.simulate('change', fakeEventDate);

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(changeFieldEvent).toHaveBeenCalledTimes(2);
    expect(changeFieldEvent).toHaveBeenNthCalledWith(1, 'margin', 4);
    expect(changeFieldEvent).toHaveBeenNthCalledWith(
      2,
      'startTime',
      '2021-05-23T11:18:21'
    );
    expect(dispatch).toHaveBeenCalledWith(Symbol.for('change'));
  });
  it('should render selects for participants', () => {
    const participants = [
      { id: '2', name: 'vlad' },
      { id: '3', name: 'vova' }
    ];
    useSelector.mockReturnValue({
      participants,
      sports: [{ id: 1, name: 'UFC' }],
      addEvent: {
        participant_Id1: '',
        participant_Id2: '',
        startTime: '2021-05-23T05:50:49',
        margin: 5
      },
      isLoading: false
    });
    const result = shallow(<CreatorEvent />);
    const options1 = result.find('.player1-option');
    const options2 = result.find('.player2-option');

    expect(options1.length).toBe(participants.length);
    expect(options2.length).toBe(participants.length);
    for (let i = 0; i < options1.length; i += 1) {
      const option = options1.at(i);
      expect(option.prop('value')).toBe(participants[i].id);
    }

    for (let i = 0; i < options2.length; i += 1) {
      const option = options2.at(i);
      expect(option.prop('value')).toBe(participants[i].id);
    }
    expect(result).toMatchSnapshot();
  });
});
