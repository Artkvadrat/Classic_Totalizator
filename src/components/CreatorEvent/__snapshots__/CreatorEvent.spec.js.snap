// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatorEvent component should render Skeleton without 1`] = `
<Skeleton
  active={true}
  avatar={false}
  paragraph={true}
  title={true}
/>
`;

exports[`CreatorEvent component should render form 1`] = `
<form
  className="form-creator"
  onSubmit={[Function]}
>
  <span
    className="sport-name"
  >
    UFC
  </span>
  <span>
    Select the first participant of event 
  </span>
  <select
    className="player-selector"
    name="participant_Id1"
    onChange={[Function]}
    value="2"
  >
    <option
      className="player1-option"
      key="2"
      value="2"
    >
      vlad
    </option>
    <option
      className="player1-option"
      key="3"
      value="3"
    >
      vova
    </option>
  </select>
  <span>
    Select the second participant of event 
  </span>
  <select
    className="player-selector"
    name="participant_Id2"
    onChange={[Function]}
    value="3"
  >
    <option
      className="player2-option"
      key="2"
      value="2"
    >
      vlad
    </option>
    <option
      className="player2-option"
      key="3"
      value="3"
    >
      vova
    </option>
  </select>
  <span>
    Enter the margin, %
  </span>
  <input
    className="margin-value"
    name="margin"
    onChange={[Function]}
    type="number"
    value={5}
  />
  <span>
    Enter the date of event
  </span>
  <input
    className="start-data"
    name="startTime"
    onChange={[Function]}
    required={true}
    type="datetime-local"
    value="2021-05-23T05:50:49"
  />
  <input
    className="create-action"
    type="submit"
    value="Create new event"
  />
</form>
`;

exports[`CreatorEvent component should render selects for partisipants 1`] = `
<form
  className="form-creator"
  onSubmit={[Function]}
>
  <span
    className="sport-name"
  >
    UFC
  </span>
  <span>
    Select the first participant of event 
  </span>
  <select
    className="player-selector"
    name="participant_Id1"
    onChange={[Function]}
    value=""
  >
    <option
      className="player1-option"
      key="2"
      value="2"
    >
      vlad
    </option>
    <option
      className="player1-option"
      key="3"
      value="3"
    >
      vova
    </option>
  </select>
  <span>
    Select the second participant of event 
  </span>
  <select
    className="player-selector"
    name="participant_Id2"
    onChange={[Function]}
    value=""
  >
    <option
      className="player2-option"
      key="2"
      value="2"
    >
      vlad
    </option>
    <option
      className="player2-option"
      key="3"
      value="3"
    >
      vova
    </option>
  </select>
  <span>
    Enter the margin, %
  </span>
  <input
    className="margin-value"
    name="margin"
    onChange={[Function]}
    type="number"
    value={5}
  />
  <span>
    Enter the date of event
  </span>
  <input
    className="start-data"
    name="startTime"
    onChange={[Function]}
    required={true}
    type="datetime-local"
    value="2021-05-23T05:50:49"
  />
  <input
    className="create-action"
    type="submit"
    value="Create new event"
  />
</form>
`;
